asllib/Interpreter.ml

```diff
@@ -1066,14 +1063,35 @@ module Make (B : Backend.S) (C : Config) = struct
 
   (* Evaluation of while and repeat loops *)
   (* ------------------------------------ *)
+
+  (* Evaluation of loop limits *)
+  and eval_loop_limit env e_limit_opt =
+    match e_limit_opt with
+    | None -> return None
+    | Some e_limit -> (
+        let* v_limit = eval_expr_sef env e_limit in
+        match B.v_to_int v_limit with
+        | Some limit -> return (Some limit)
+        | None ->
+            fatal_from e_limit
+              (Error.MismatchType (B.debug_value v_limit, [ integer' ])))
+
+  and tick_loop_limit loc limit_opt =
+    match limit_opt with
+    | None -> return None
+    | Some limit ->
+        if limit >= 1 then return (Some (limit - 1))
+        else fatal_from loc Error.LoopLimitReached
+
   (* Begin EvalLoop *)
-  and eval_loop is_while env e_cond body : stmt_eval_type =
+  and eval_loop loc is_while env limit_opt e_cond body : stmt_eval_type =
     (* Name for warn messages. *)
     let loop_name = if is_while then "While loop" else "Repeat loop" in
     (* Continuation in the positive case. *)
     let loop env =
+      let* limit_opt' = tick_loop_limit loc limit_opt in
```

Comment:
perhap switch these lines for clarity (decrementing the loop limit after an iteration).